modules/apps/transfer/keeper/keeper.go

```diff
@@ -133,14 +137,83 @@ func (k Keeper) IterateDenomTraces(ctx sdk.Context, cb func(denomTrace types.Den
 
 	defer sdk.LogDeferred(ctx.Logger(), func() error { return iterator.Close() })
 	for ; iterator.Valid(); iterator.Next() {
-
 		denomTrace := k.MustUnmarshalDenomTrace(iterator.Value())
 		if cb(denomTrace) {
 			break
 		}
 	}
 }
 
+// GetTotalEscrowForDenom gets the total amount of source chain tokens that are in escrow.
+func (k Keeper) GetTotalEscrowForDenom(ctx sdk.Context, denom string) math.Int {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get(types.TotalEscrowForDenomKey(denom))
+	if bz == nil {
+		return math.ZeroInt()
+	}
+
+	var amount math.Int
+	if err := amount.Unmarshal(bz); err != nil {
+		panic(err)
+	}
+	return amount
+}
+
+// SetTotalEscrowForDenom stores the total amount of source chain tokens that are in escrow.
+func (k Keeper) SetTotalEscrowForDenom(ctx sdk.Context, denom string, amount math.Int) {
+	if amount.LT(math.ZeroInt()) {
+		panic(fmt.Sprintf("amount cannot be negative: %s", amount))
+	}
+
+	store := ctx.KVStore(k.storeKey)
+	bz, err := amount.Marshal()
+	if err != nil {
+		panic(err)
+	}
+
+	store.Set(types.TotalEscrowForDenomKey(denom), bz)
+}
+
+// GetAllDenomEscrows returns the escrow information for all the denominations.
+func (k Keeper) GetAllDenomEscrows(ctx sdk.Context) sdk.Coins {
+	escrows := sdk.Coins{}
+	k.IterateDenomEscrows(ctx, func(denomEscrow sdk.Coin) bool {
+		escrows = append(escrows, denomEscrow)
+		return false
+	})
+
+	return escrows.Sort()
+}
+
+// IterateDenomEscrows iterates over the denomination escrows in the store
+// and performs a callback function.
+func (k Keeper) IterateDenomEscrows(ctx sdk.Context, cb func(denomEscrow sdk.Coin) bool) {
+	re := regexp.MustCompile(fmt.Sprintf(`%s\/%s\/(.*[^\s])`, types.KeyTotalEscrowPrefix, types.KeyDenomsPrefix))
+
+	store := ctx.KVStore(k.storeKey)
+	iterator := sdk.KVStorePrefixIterator(store, []byte(types.KeyTotalEscrowPrefix))
+
+	defer sdk.LogDeferred(ctx.Logger(), func() error { return iterator.Close() })
+	for ; iterator.Valid(); iterator.Next() {
+		matches := re.FindStringSubmatch(string(iterator.Key()))
+		if len(matches) != 2 { // there should be two matches: 1st one for the whole string, 2nd for the denomination
+			continue
+		}
```

Comment:
An alternative is to use [`strings.Contains`](https://github.com/cosmos/ibc-go/blob/main/modules/core/02-client/keeper/keeper.go#L376). My main gripe with regexp is the `MustCompile` statement is hard to read, I wonder if we should keep things simple?

Reply:
I share a similar feeling about regular expression :) but I think they also simplify the code quite a bit.